Feature/player stats panel

Rheyzze opened the pull request.

Rheyzze committed:
feat: create panel with player info and result

Rheyzze committed:
feat:rota

Rheyzze committed:
feat: add start screen and buttons to reset match and players

italobarrosme commented on src/modules/games/tictactoe/hook/useTicTacToe.tsx:
```diff
@@ -34,6 +34,7 @@ export const useTicTacToe = () => {
   const [gameState, setGameState] = useState<GameState>('playing')
   const [winner, setWinner] = useState<Player | null>(null)
   const [winningLine, setWinningLine] = useState<number[] | null>(null)
+  const [players, setPlayers] = useState({ x: 'Jogador 1', o: 'Jogador 2' })
```
Não há necessidade de incluir os players na lógica desse custom hook. Eles apenas utilizam essa lógica, mas não precisam estar diretamente contidos nela.

italobarrosme commented on src/modules/games/tictactoe/hook/index.ts:
```diff
@@ -1 +1 @@
-export * from './useTictactoe'
```
Boa correção!

italobarrosme commented on src/modules/games/tictactoe/components/StartScreen/index.ts:
```diff
@@ -0,0 +1 @@
+export { default as StartScreen } from './StartScreen'
```
como padrão vamos exportar nomeado se alias

italobarrosme commented on src/modules/games/tictactoe/components/StartScreen/StartScreen.tsx:
```diff
@@ -0,0 +1,50 @@
+  onStart: (players: { p1: string; p2: string }) => void
+}
+
+export default function StartScreen({ onStart }: Props) {
```
esse export não precisar ser default pode ser nomeado

italobarrosme replied:
esse export não precisar ser default pode ser nomeado

italobarrosme commented on src/modules/games/tictactoe/components/StartScreen/StartScreen.tsx:
```diff
@@ -0,0 +1,50 @@
+}
+
+export default function StartScreen({ onStart }: Props) {
+  const [player1, setPlayer1] = useState('')
```
Como mencionado em uma das tasks sobre os Ranks utilizando o Zustand, já que iremos persistir os dados do jogador, o ideal é criar uma store dedicada para o player. Dessa forma, utilizamos essa store diretamente, sem a necessidade de usar o hook de state do React.

italobarrosme commented on src/modules/games/tictactoe/components/StartScreen/StartScreen.tsx:
```diff
@@ -0,0 +1,50 @@
+  const [player2, setPlayer2] = useState('')
+
+  const handleStart = () => {
+    if (!player1.trim() || !player2.trim()) return
```
Boa logica, porem quando o texto for persistido na store esse trim() pode ir na hora que a store receber o dado

italobarrosme commented on src/modules/games/tictactoe/components/StartScreen/StartScreen.tsx:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  return (
+    <div className="flex items-center justify-center min-h-screen bg-[#0a003d] text-white font-['Press_Start_2P']">
```
Sugiro não usar o bg-[#0a003d] font-['Press_Start_2P'] dessa forma é hardcoded, implemente essa cor e fonte na configuração do thema do projeto no arquivo style

italobarrosme commented on src/modules/games/tictactoe/components/StartScreen/StartScreen.tsx:
```diff
@@ -0,0 +1,50 @@
+  return (
+    <div className="flex items-center justify-center min-h-screen bg-[#0a003d] text-white font-['Press_Start_2P']">
+      <div className="text-center space-y-6">
+        <h1 className="text-3xl text-yellow-400 drop-shadow-[2px_2px_0px_rgba(0,0,0,1)]">
```
mesma coisa aqui

italobarrosme commented on src/modules/games/tictactoe/components/StartScreen/StartScreen.tsx:
```diff
@@ -0,0 +1,50 @@
+        </h1>
+
+        <div className="space-y-4">
+          <input
```
Esse formulário está funcionando, mas existem algumas sintaxes padrão que podem ser aplicadas. Por exemplo, usar a tag `<form>`

Com isso, você ganha vários pontos positivos: semântica, suporte nativo à submissão do formulário, eventos integrados e até validações nativas caso queira aproveitar. Além disso, facilita o tratamento de mensagens de erro.

italobarrosme commented on src/modules/games/tictactoe/components/StartScreen/StartScreen.tsx:
```diff
@@ -0,0 +1,50 @@
+        </div>
+
+        <button
+          onClick={handleStart}
```
usando o form esse `button `vira type submit

italobarrosme commented on src/modules/games/tictactoe/components/GameInfoPanel/GameInfoPanel.tsx:
```diff
@@ -0,0 +1,53 @@
+      <div className="text-lg">
+        {gameState === 'playing' && (
+          <p>
+            Vez de:{' '}
```
existem outras formas de da espaço com uma escrita melhor usando` template string`

italobarrosme commented on src/modules/games/tictactoe/components/GameInfoPanel/GameInfoPanel.tsx:
```diff
@@ -0,0 +1,53 @@
+        {gameState === 'playing' && (
+          <p>
+            Vez de:{' '}
+            {players[currentPlayer.toLowerCase() as keyof typeof players]}
```
Não precisa desse alias `currentPlayer.toLowerCase() as keyof typeof players`
Dei uma fuçada pra entender qual era o error, mas eu vi aqui voce ta atribuindo 

players: { x: string; o: string } com o minusculo, acredito que ele começou como maisculo por isso esse error corrija em todos os typers que atribua o objeto `o:` pra maisculo ou minuscolo

italobarrosme commented on src/modules/games/tictactoe/components/GameInfoPanel/GameInfoPanel.tsx:
```diff
@@ -0,0 +1,53 @@
+        )}
+        {isDraw && <p>O jogo empatou!</p>}
+      </div>
+      {(gameState === 'won' || isDraw) && (
```
Sobre os estados dos botoes pode da um return de acordo com a condicional gerando menos codigo

italobarrosme commented on src/modules/3d/Scene/Scene.tsx:
```diff
@@ -8,15 +8,18 @@ import { Block } from '@/modules/games/tictactoe/3d/Block'
+  blocksPositions: ReturnType<typeof useTicTacToe>['blocksPositions']
+}
 
+export const Scene = ({
```
Podemos ter um jeito melhor de lidar com o form sem passar via props pra Scene

italobarrosme commented on src/app/test/page.tsx:
```diff
@@ -0,0 +1,10 @@
+'use client'
+import { StartScreen } from '@/modules/games/tictactoe/components/StartScreen'
+
+export default function TestPage() {
```
esse arquivo ta sendo usado? se não remova-o

italobarrosme commented on src/app/page.tsx:
```diff
@@ -1,14 +1,5 @@
-  }, [])
-
-  return null
+  redirect('/games')
```
Nice correção muito bom!

italobarrosme commented on src/app/games/page.tsx:
```diff
@@ -10,9 +14,56 @@ const RenderCanvas = dynamic(() => import('@/modules/3d/RenderCanvas'), {
 })
 
 export default function Game() {
+  const [gameStarted, setGameStarted] = useState(false)
```
essa logica aqui na page, não pode acontecer o game started pode ser feito de outra forma, talvez retornando da logica do hook do useTicTacToe

italobarrosme reviewed (changes requested):
De um olhada nos comentarios, e veja os pontos de melhoria

Rheyzze committed:
feat(tictactoe): refactor game logic to use zustand stores and improve components

refactor: move game state to zustand stores for better state management
feat: add PixelButton component for consistent UI
fix: correct X component rotation in 3D view
style: update global CSS with game-specific styles
test: update test mocks to match new store structure
chore: update zustand dependency to latest version

italobarrosme commented on package.json:
```diff
@@ -51,7 +51,7 @@
     "usehooks-ts": "^3.1.1",
     "zod": "^3.25.63",
-    "zustand": "^5.0.5"
+    "zustand": "^5.0.8"
```
Teve algum motivo especifico para atualizar a lib do Zustand?
ps: é delicado atualizar libs deve haver uma boa comunicacao entre o time sobre esse assunto.

italobarrosme commented on src/app/games/page.spec.tsx:
```diff
@@ -0,0 +1,79 @@
+import { render, screen, fireEvent, waitFor } from '@testing-library/react'
```
Sobre os arquivos de testes a gente tem que trocar uma idea. pra que quando for feito com AI, ela siga o padrão do projeto e fique facil a leitura do codigo. a principio vamos deletar esses testes e deixar para cobrir em um outro momento

italobarrosme commented on src/modules/3d/RenderCanvas/RenderCanvas.tsx:
```diff
@@ -14,6 +15,23 @@ export const RenderCanvas = ({ children }: RenderCanvasProps) => {
+      }
+    }
+
+    window.addEventListener('reset-camera-position', handleResetCamera)
```
Um otimo pensamento, mas acredito que podemos seguir um padrao tirando essa dependencia de usar o obejto window

italobarrosme commented on src/stores/ticTacToeStore.ts:
```diff
@@ -0,0 +1,100 @@
+import { create } from 'zustand'
```
a store era so pra lidar com os players, tu acabou jogando a logica do custom hook pra dentro da store, esse nao era o caminho

italobarrosme committed:
refactor: simplify PixelButton component by using utility function for class names

italobarrosme committed:
refactor: reorganize game components and state management for tic-tac-toe

italobarrosme committed:
refactor: remove obsolete test file for game page

italobarrosme committed:
feat(tictactoe): implement start screen and game info panel updates

Rheyzze committed:
feat(tictactoe): integrate useTicTacToe hook into Game and GameInfoPanel components

Rheyzze committed:
refactor: format Scene component props for improved readability

Rheyzze committed:
feat(tictactoe): refactor useTicTacToe hook to use unified state object and localStorage for persistence

Rheyzze committed:
refactor(useTicTacToe): improve readability by formatting callback parameters and state updates

Rheyzze committed:
fix(generateUniqueId): ensure performance.now() is safely accessed in non-browser environments

Rheyzze committed:
refactor(generateUniqueId): improve readability by formatting performanceNow assignment

Rheyzze committed:
refactor(index): change export statement for useTicTacToe to named export for clarity

italobarrosme commented on src/modules/3d/Scene/Scene.tsx:
```diff
@@ -1,25 +1,22 @@
-export const Scene = () => {
-  const { blocksPositions, board, handleBlockClick } = useTicTacToe()
+// 1. Definir as props da cena com base no que é retornado pelo hook useTicTacToe
+type SceneProps = ReturnType<typeof useTicTacToe>
```
tipe a scena de acordo com uso dela, usar a tipagem do hook vai estar atribuindo dados que nao fazem parte do componente Scene, exemplo abaixo

```
type SceneProps = {
  blocksPositions: readonly {
    position: readonly [number, number, number]
    color: string
  }[]
  board: Array<'normal' | 'x' | 'o'>
  handleBlockClick: (index: number) => void
}
```

italobarrosme commented on src/modules/games/tictactoe/components/GameInfoPanel/GameInfoPanel.tsx:
```diff
@@ -0,0 +1,83 @@
+    redirect('/')
+  }
+
+  return (
```
isso pode ser simplificado dividindo em 2 componentes exemplo:


```
'use client'

import { GAME_INFO_PANEL_TEXTS } from './constants'
import { GAME_STATE } from '../../constants/game'
import { usePlayerStore } from '../../store'
import { redirect } from 'next/navigation'

type GameInfoPanelProps = {
  gameState: 'playing' | 'won' | 'draw'
  currentPlayer: 'x' | 'o'
  winner: 'x' | 'o' | null
  isDraw: boolean
  resetGame: () => void
}

export const GameInfoPanel = ({
  gameState,
  currentPlayer,
  winner,
  isDraw,
  resetGame,
}: GameInfoPanelProps) => {
  const { players, resetPlayers } = usePlayerStore()

  const onRestGame = () => {
    resetGame()
  }

  const onResetPlayers = () => {
    resetPlayers()
    resetGame()
    redirect('/')
  }

  return (
    <div className="absolute top-4 left-4 bg-white p-4 rounded-lg shadow-lg text-gray-800 z-10">
      <h2 className="text-xl font-bold mb-2">{GAME_INFO_PANEL_TEXTS.title}</h2>
      <div className="text-lg">
        {gameState === GAME_STATE.PLAYING && (
          <p>{`${GAME_INFO_PANEL_TEXTS.turn} ${players[currentPlayer]}`}</p>
        )}
        {gameState === GAME_STATE.WON && winner && (
          <p>{`${players[winner]} ${GAME_INFO_PANEL_TEXTS.winner}`}</p>
        )}
        {isDraw && <p>{GAME_INFO_PANEL_TEXTS.draw}</p>}
      </div>
      <div>
        {GameInfoPanelController({
          onRestGame,
          onResetPlayers,
          gameState,
          isDraw,
        })}
      </div>
    </div>
  )
}

type GameInfoPanelControllerProps = {
  onRestGame: () => void
  onResetPlayers: () => void
  gameState: 'playing' | 'won' | 'draw'
  isDraw: boolean
}

const GameInfoPanelController = ({
  onRestGame,
  onResetPlayers,
  gameState,
  isDraw,
}: GameInfoPanelControllerProps) => {
  if (gameState === GAME_STATE.WON || isDraw) {
    return (
      <>
        <button
          onClick={onRestGame}
          className="mt-4 px-4 py-2 bg-blue-500 text-white rounded hover:bg-blue-600"
        >
          {GAME_INFO_PANEL_TEXTS.restartRound}
        </button>
        <button
          onClick={onResetPlayers}
          className="mt-4 ml-2 px-4 py-2 bg-gray-500 text-white rounded hover:bg-gray-600"
        >
          {GAME_INFO_PANEL_TEXTS.restartPlayers}
        </button>
      </>
    )
  }

  return (
    <button
      onClick={onResetPlayers}
      className="mt-4 px-4 py-2 bg-gray-500 text-white rounded hover:bg-gray-600"
    >
      {GAME_INFO_PANEL_TEXTS.restartPlayers}
    </button>
  )
}

```

italobarrosme commented on src/modules/games/tictactoe/components/GameInfoPanel/GameInfoPanel.tsx:
```diff
@@ -0,0 +1,83 @@
+import { useTicTacToe } from '../../hook'
+
+// Tipagem D.R.Y: Os tipos são derivados diretamente do hook.
+type GameState = ReturnType<typeof useTicTacToe>
```
ta redundante essa tipagem

italobarrosme commented on src/modules/games/tictactoe/components/GameInfoPanel/GameInfoPanel.tsx:
```diff
@@ -0,0 +1,83 @@
+  resetGame,
+}: GameInfoPanelProps) => {
+  const { players, resetPlayers } = usePlayerStore()
+  const [isClient, setIsClient] = useState(false)
```
Nao sei se precisa desse estado nao, verifica o meu exemplo

italobarrosme commented on src/modules/games/tictactoe/hook/useTicTacToe.tsx:
```diff
@@ -28,96 +37,131 @@ const winningCombinations = [
 ]
 
+// Função que define o estado inicial, lendo do localStorage se possível
+const getInitialState = (): GameState => {
```
use o Zustand pra fazer isso

italobarrosme commented on src/modules/games/tictactoe/hook/useTicTacToe.tsx:
```diff
@@ -28,96 +37,131 @@ const winningCombinations = [
+
+  // 2. Este useEffect salva o estado no localStorage sempre que ele muda
+  useEffect(() => {
+    localStorage.setItem('ticTacToeGameState', JSON.stringify(state))
```
a gente esta persistindo no zustand nao sei se precisa disso nao!

italobarrosme commented on src/modules/games/tictactoe/hook/useTicTacToe.tsx:
```diff
@@ -28,96 +37,131 @@ const winningCombinations = [
+}
+
 export const useTicTacToe = () => {
-  const [board, setBoard] = useState<Array<CellState>>(Array(9).fill('normal'))
```
removeu todos os states do logica do hook, axo q isso aqui seria o caminho se pegarmos o objeto vindo do zustand

italobarrosme commented on src/modules/games/tictactoe/components/StartScreen/StartScreen.tsx:
```diff
@@ -0,0 +1,77 @@
+  const { players, setPlayer } = usePlayerStore()
+  const { resetGame } = useTicTacToe()
+
+  // useEffect(() => {
```
se nao ta usando remove

italobarrosme commented on src/modules/games/tictactoe/hook/index.ts:
```diff
@@ -1 +1 @@
-export * from './useTictactoe'
```
teve algum motivo pra mudar isso?

Rheyzze replied:
tava com sono kkkkkkk, ai tava olhando, e mudei sem motivo mesmo xD

italobarrosme commented on src/app/games/page.tsx:
```diff
@@ -10,9 +13,20 @@ const RenderCanvas = dynamic(() => import('@/modules/3d/RenderCanvas'), {
+        resetGame={game.resetGame}
+      />
+      <RenderCanvas>
+        <Scene {...game} />
```
Quando um componente recebe poucas propriedades (props), manter as nomeações explícitas realmente é a melhor prática

Rheyzze committed:
refactor(tictactoe): restructure game state management using Zustand for improved performance and maintainability
